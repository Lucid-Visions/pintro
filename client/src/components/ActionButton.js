import React from 'react';
import { View, StyleSheet, Text, TouchableOpacity } from 'react-native';
import ElevatedView from 'react-native-elevated-view'
import PropTypes from 'prop-types';
import { Ionicons } from '@expo/vector-icons';

/**
 * Renders an action button for the user's profile.
 * @param {string} type the type of the button.
 * @param {string} context the specific topic related to the type. 
 */
const ActionButtonComponent = ({ type, context }) => (

    <ElevatedView style={styles.view} elevation={10}>
        <TouchableOpacity style={styles.container}>
            <View style={styles.text}>
                <Text style={styles.title}>{resolveType(type)}</Text>
                <Text style={styles.subtitle}>{context}</Text>
            </View>
            <View style={styles.outline}>
                <View style={styles.icon}>
                    <Ionicons name="ios-chatbubbles" size={32} />
                </View>
            </View>
        </TouchableOpacity>
        </ElevatedView>
);

/**
 * Displays text according to the input type.
 * @param {string} type the type of the button.
 */
function resolveType(type) {
    switch (type) {
        case "help": return "Help me with";
        case "introduce": return "Introduce me to";
        case "talk": return "Talk to me about";
    }
}

/**
 * Custom validator for the type input.
 */
const typeValidator = (props, propName, componentName) => {
    let component = componentName || 'ANONYMOUS';
    let validTypes = ["help", "introduce", "talk"];

    if (props[propName]) {
        let value = props[propName];
        return validTypes.includes(value)
            ? null
            : new Error(`${value} is not a valid type! Only ${validTypes.toString()} are permitted.`);
    }

    //Assume everything is ok
    return null;
}

// Type checker
ActionButtonComponent.propTypes = {
    type: typeValidator,
    context: PropTypes.string,
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: 'row',
    },
    view: {
        alignItems: 'center',
        flexDirection: 'row',
        backgroundColor: 'white',
        borderRadius: 15,
        width: 350,
        height: '10%',
        padding: 20,
        minHeight: 100,
        margin: 4,
    },
    text: {
        flexDirection: 'column',
        left: 'auto',
    },
    title: {
        fontWeight: 'bold',
        fontSize: 22
    },
    subtitle: {
        fontWeight: 'normal',
        fontSize: 18,
        color: 'grey'
    },
    outline: {
        alignItems: 'center',
        justifyContent: 'center',
        marginTop:-5,
        flexDirection: 'row',
        right:20,
        position:"absolute",
        padding: 2,
        borderWidth: 2,
        borderColor: 'orange',
        borderRadius: 150 / 2
    },
    icon: {
        alignItems: 'center',
        justifyContent: 'center',
        flexDirection: 'row',
        marginLeft: 'auto',
        backgroundColor: 'orange',
        width: 50,
        height: 50,
        borderRadius: 50 / 2,

    }
});

export default ActionButtonComponent;