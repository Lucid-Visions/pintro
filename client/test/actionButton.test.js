import React from "react";
import ActionButton from "../src/components/ActionButton";
import renderer from "react-test-renderer";

/**
 * Test the action buttons in the personal profile, 
 * this component doesn't have an onPress method implemented, so we 
 * cannot test if it actually works on press, we check if it renders properly,
 * and if the props type checking is working as expected.
 */
describe("<ActionButton />", () => {
  test("help me with renders correctly", () => {
    const buttonTree = renderer
      .create(<ActionButton type="help" context="Unit Testing" />)
      .toJSON();
    expect(buttonTree).toMatchSnapshot();
  });

  test("introduce me to render correctly", () => {
    const buttonTree = renderer
      .create(<ActionButton type="introduce" context="Bill Gates" />)
      .toJSON();
    expect(buttonTree).toMatchSnapshot();
  });

  test("talk to me to render correctly", () => {
    const buttonTree = renderer
      .create(<ActionButton type="talk" context="Tic Tac Toe" />)
      .toJSON();
    expect(buttonTree).toMatchSnapshot();
  });

  test("action button with wrong type, should return warning to console", () => {
    var warn = jest.spyOn(global.console, "warn");

    const buttonTree = renderer.create(<ActionButton type="noType" />).toJSON();

    expect(warn).not.toHaveBeenCalled();

    // Cleanup
    warn.mockReset();
    warn.mockRestore();
  });
});
