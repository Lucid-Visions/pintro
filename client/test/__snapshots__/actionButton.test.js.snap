// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionButton /> help me with renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "shadowOffset": Object {
          "height": 6,
        },
        "shadowOpacity": 0.195,
        "shadowRadius": 5.4,
      },
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 15,
        "flexDirection": "row",
        "height": "10%",
        "margin": 4,
        "minHeight": 100,
        "padding": 20,
        "width": 350,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "left": "auto",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 22,
            "fontWeight": "bold",
          }
        }
      >
        Help me with
      </Text>
      <Text
        style={
          Object {
            "color": "grey",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        Unit Testing
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "borderColor": "orange",
          "borderRadius": 75,
          "borderWidth": 2,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": -5,
          "padding": 2,
          "position": "absolute",
          "right": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "orange",
            "borderRadius": 25,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "marginLeft": "auto",
            "width": 50,
          }
        }
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;

exports[`<ActionButton /> introduce me to render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "shadowOffset": Object {
          "height": 6,
        },
        "shadowOpacity": 0.195,
        "shadowRadius": 5.4,
      },
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 15,
        "flexDirection": "row",
        "height": "10%",
        "margin": 4,
        "minHeight": 100,
        "padding": 20,
        "width": 350,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "left": "auto",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 22,
            "fontWeight": "bold",
          }
        }
      >
        Introduce me to
      </Text>
      <Text
        style={
          Object {
            "color": "grey",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        Bill Gates
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "borderColor": "orange",
          "borderRadius": 75,
          "borderWidth": 2,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": -5,
          "padding": 2,
          "position": "absolute",
          "right": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "orange",
            "borderRadius": 25,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "marginLeft": "auto",
            "width": 50,
          }
        }
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;

exports[`<ActionButton /> talk to me to render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "shadowOffset": Object {
          "height": 6,
        },
        "shadowOpacity": 0.195,
        "shadowRadius": 5.4,
      },
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 15,
        "flexDirection": "row",
        "height": "10%",
        "margin": 4,
        "minHeight": 100,
        "padding": 20,
        "width": 350,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "left": "auto",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 22,
            "fontWeight": "bold",
          }
        }
      >
        Talk to me about
      </Text>
      <Text
        style={
          Object {
            "color": "grey",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        Tic Tac Toe
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "borderColor": "orange",
          "borderRadius": 75,
          "borderWidth": 2,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": -5,
          "padding": 2,
          "position": "absolute",
          "right": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "orange",
            "borderRadius": 25,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "marginLeft": "auto",
            "width": 50,
          }
        }
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;
